### ISSUE TITLE
Ubuntu 23.10: cloud-init status --wait blocks indefinitely on cloud-config.service when APT keys in user-data

### ISSUE THREAD

**Issue by blackboxsw** [canonical/cloud-init#4344]
# Bug report
cloud-init status --wait never returns due to systemd SubState=running for cloud-config.service whenever user-data includes apt keyserver/keyid

Distribution/Release: Ubuntu 23.10 (Mantic) 
Systemd version: 253.5-1ubuntu1
cloud-init version: 99.daily-202308050431-441d8f81~ubuntu23.10.1
APT version: apt 2.6.0

Something has changed in how systemd or APT is behaving on Ubuntu Mantic which triggers the SubState of cloud-config.service in Ubuntu 23.10 to report SubState=running instead of 'exited'.  This 'SubState=running'  prevents cloud-init status --wait from returning which blocks any external scripts or tools from being run resulting in integration test failures or end-users being unable to determine when it is "safe" to manage the VM without cloud-init ongoing configuration collisions.

Lunar  images do not exhibit this problem with the same version of cloud-init as on Mantic

* Mantic config details showing cloud-config.service as running
```
root@mantic:~# systemd --version
systemd 253 (253.5-1ubuntu1)
+PAM +AUDIT +SELINUX +APPARMOR +IMA +SMACK +SECCOMP +GCRYPT -GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY +P11KIT +QRENCODE +TPM2 +BZIP2 +LZ4 +XZ +ZLIB +ZSTD -BPF_FRAMEWORK -XKBCOMMON +UTMP +SYSVINIT default-hierarchy=unified
root@mantic:~# apt --version
apt 2.7.2 (amd64)
root@mantic:~# cloud-init --version
/usr/bin/cloud-init 99.daily-202308141631-4a37eaaa~ubuntu23.10.1
root@mantic:~# systemctl status cloud-config.service
● cloud-config.service - Apply the settings specified in cloud-config
     Loaded: loaded (/lib/systemd/system/cloud-config.service; enabled; preset: enabled)
     Active: active (running) since Mon 2023-08-14 17:27:58 UTC; 29s ago
    Process: 403 ExecStart=/usr/bin/cloud-init modules --mode=config (code=exited, status=0/SUCCESS)
   Main PID: 403 (code=exited, status=0/SUCCESS)
      Tasks: 2 (limit: 18656)
     Memory: 23.4M
        CPU: 2.449s
     CGroup: /system.slice/cloud-config.service
             ├─447 dirmngr --daemon --homedir /root/.gnupg
             └─453 gpg-agent --homedir /root/.gnupg --use-standard-socket --daemon

Aug 14 17:27:57 mantic cloud-init[416]: W: Target CNF (main/cnf/Commands-amd64) is configured multiple times in /etc/apt/sources.list.d/cloudinit.list:1 and /etc/apt/sources.list.d/c>
Aug 14 17:27:57 mantic cloud-init[416]: W: Target CNF (main/cnf/Commands-all) is configured multiple times in /etc/apt/sources.list.d/cloudinit.list:1 and /etc/apt/sources.list.d/clo>
Aug 14 17:27:57 mantic cloud-init[416]: W: Target Packages (main/binary-amd64/Packages) is configured multiple times in /etc/apt/sources.list.d/cloudinit.list:1 and /etc/apt/sources.>
Aug 14 17:27:57 mantic cloud-init[416]: W: Target Packages (main/binary-all/Packages) is configured multiple times in /etc/apt/sources.list.d/cloudinit.list:1 and /etc/apt/sources.li>
Aug 14 17:27:57 mantic cloud-init[416]: W: Target Translations (main/i18n/Translation-en) is configured multiple times in /etc/apt/sources.list.d/cloudinit.list:1 and /etc/apt/source>
Aug 14 17:27:57 mantic cloud-init[416]: W: Target DEP-11 (main/dep11/Components-amd64.yml) is configured multiple times in /etc/apt/sources.list.d/cloudinit.list:1 and /etc/apt/sourc>
Aug 14 17:27:57 mantic cloud-init[416]: W: Target DEP-11 (main/dep11/Components-all.yml) is configured multiple times in /etc/apt/sources.list.d/cloudinit.list:1 and /etc/apt/sources>
Aug 14 17:27:58 mantic cloud-init[416]: W: Target CNF (main/cnf/Commands-amd64) is configured multiple times in /etc/apt/sources.list.d/cloudinit.list:1 and /etc/apt/sources.list.d/c>
Aug 14 17:27:58 mantic cloud-init[416]: W: Target CNF (main/cnf/Commands-all) is configured multiple times in /etc/apt/sources.list.d/cloudinit.list:1 and /etc/apt/sources.list.d/clo>
Aug 14 17:27:58 mantic systemd[1]: Finished cloud-config.service - Apply the settings specified in cloud-config.
root@mantic:~# systemctl show  --property=ActiveState,UnitFileState,SubState
root@mantic:~# systemctl show  --property=ActiveState,UnitFileState,SubState cloud-config.service
ActiveState=active
SubState=running
UnitFileState=enabled
```

* Lunar config details showing cloud-config.service as exited
```
root@lunar:~# systemd --version: 
systemd 252 (252.5-2ubuntu3)
+PAM +AUDIT +SELINUX +APPARMOR +IMA +SMACK +SECCOMP +GCRYPT -GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY +P11KIT +QRENCODE +TPM2 +BZIP2 +LZ4 +XZ +ZLIB +ZSTD -BPF_FRAMEWORK -XKBCOMMON +UTMP +SYSVINIT default-hierarchy=unified
root@clean:~# apt --version
apt 2.6.0 (amd64)
root@lunar:~# gpg-agent --version
gpg-agent (GnuPG) 2.2.40
libgcrypt 1.10.1
Copyright (C) 2022 g10 Code GmbH
License GNU GPL-3.0-or-later <https://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
root@clean:~# systemctl show cloud-config.service --property=ActiveState,UnitFileState,SubState
ActiveState=active
SubState=exited
UnitFileState=enabled
root@me:~# systemctl status cloud-config.service
● cloud-config.service - Apply the settings specified in cloud-config
     Loaded: loaded (/lib/systemd/system/cloud-config.service; enabled; preset: enabled)
     Active: active (exited) since Mon 2023-08-14 17:16:23 UTC; 8min ago
    Process: 389 ExecStart=/usr/bin/cloud-init modules --mode=config (code=exited, status=0/SUCCESS)
   Main PID: 389 (code=exited, status=0/SUCCESS)
      Tasks: 2 (limit: 18656)
     Memory: 21.8M
        CPU: 2.382s
     CGroup: /system.slice/cloud-config.service
             ├─455 dirmngr --daemon --homedir /root/.gnupg
             └─461 gpg-agent --homedir /root/.gnupg --use-standard-socket --daemon
root@me:~# systemctl show cloud-config.service --property=ActiveState,UnitFileState,SubState
ActiveState=active
SubState=exited
UnitFileState=enabled
```

## Steps to reproduce the problem

```
cat << EOF > systemd-apt-hanging-config.yaml
#cloud-config
apt:
  sources:
    cloudinit: 
      source: 'deb [arch=amd64] http://ppa.launchpad.net/cloud-init-dev/daily/ubuntu mantic main'
      keyserver: keyserver.ubuntu.com
      keyid: E4D304DF
EOF

lxc launch ubuntu-daily:mantic config-hang -c cloud-init.user-data="$(cat systemd-apt-hanging-config.yaml)"
lxc exec config-hang -- cloud-init status --long --wait
```

**Comment by holmanb** [comment]
odd, it looks like cloud-config completes:
```
2023-08-14 17:42:59,245 - util.py[DEBUG]: cloud-init mode 'modules' took 11.084 seconds (11.09)
2023-08-14 17:42:59,245 - handlers.py[DEBUG]: finish: modules-config: SUCCESS: running modules for config
2023-08-14 17:42:59,632 - util.py[DEBUG]: Cloud-init v. 23.3~3gebd7b2b6-0ubuntu1 running 'modules:final' at Mon, 14 Aug 2023 17:42:59 +0000. Up 21.71 seconds.
```

**Comment by TheRealFalcon** [comment]
Wow, we even get the systemd message that the service is finishing...
```
root@me:~# systemctl status cloud-config.service
● cloud-config.service - Apply the settings specified in cloud-config
     Loaded: loaded (/lib/systemd/system/cloud-config.service; enabled; preset: enabled)
     Active: active (running) since Wed 2023-08-16 01:45:48 UTC; 1min 59s ago
    Process: 1031 ExecStart=/usr/bin/cloud-init modules --mode=config (code=exited, status=0/SUCCESS)
   Main PID: 1031 (code=exited, status=0/SUCCESS)
      Tasks: 2 (limit: 76591)
     Memory: 246.2M
        CPU: 5.349s
     CGroup: /system.slice/cloud-config.service
             ├─1058 dirmngr --daemon --homedir /root/.gnupg
             └─1078 gpg-agent --homedir /root/.gnupg --use-standard-socket --daemon

Aug 16 01:45:45 me cloud-init[1034]: Get:21 http://archive.ubuntu.com/ubuntu mantic/multiverse amd64 c-n-f Metadata [8832 B]
Aug 16 01:45:45 me cloud-init[1034]: Get:22 http://archive.ubuntu.com/ubuntu mantic-updates/universe amd64 c-n-f Metadata [112 B]
Aug 16 01:45:45 me cloud-init[1034]: Get:23 http://archive.ubuntu.com/ubuntu mantic-updates/multiverse amd64 c-n-f Metadata [116 B]
Aug 16 01:45:45 me cloud-init[1034]: Get:24 http://archive.ubuntu.com/ubuntu mantic-backports/main amd64 c-n-f Metadata [112 B]
Aug 16 01:45:45 me cloud-init[1034]: Get:25 http://archive.ubuntu.com/ubuntu mantic-backports/restricted amd64 c-n-f Metadata [116 B]
Aug 16 01:45:45 me cloud-init[1034]: Get:26 http://archive.ubuntu.com/ubuntu mantic-backports/universe amd64 c-n-f Metadata [116 B]
Aug 16 01:45:45 me cloud-init[1034]: Get:27 http://archive.ubuntu.com/ubuntu mantic-backports/multiverse amd64 c-n-f Metadata [116 B]
Aug 16 01:45:48 me cloud-init[1034]: Fetched 28.0 MB in 4s (6699 kB/s)
Aug 16 01:45:48 me cloud-init[1034]: Reading package lists...
Aug 16 01:45:48 me systemd[1]: Finished cloud-config.service - Apply the settings specified in cloud-config.
```

**Comment by TheRealFalcon** [comment]
This is being caused because the gpg call(s) is spawning two new daemons:
```
     CGroup: /system.slice/cloud-config.service
             ├─1058 dirmngr --daemon --homedir /root/.gnupg
             └─1078 gpg-agent --homedir /root/.gnupg --use-standard-socket --daemon
```

One possible solution is to run  `gpgconf --kill all` after all other `gpg` calls have been run.

**Comment by blackboxsw** [comment]
Ok so, newer systemd v.253 tracks SubState based on any spawned daemons in the same CGroup. This behavior doesn't exist on systemd 252 and earlier. If those daemons are still running and active, then the owning systemd service is reported as still `SubState=running` instead of `SubState=exited`.

Will have to dig into the systemd implementation to see if an alternative when spawning daemons is desirable, but, in the case of apt setting up keys, we probably don't want gpg-agent and dirmgr to live after cloud-config.service has run anyway. So Jame's gpgconfg --kill all makes sense in early boot for this scenario

**Comment by blackboxsw** [comment]
adding priority label to indicate we want this fix for 23.3
